google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java

```diff
@@ -1306,6 +1315,93 @@ public void deleteNestedFieldUsingFieldPath() throws Exception {
+                ref.compareAndSet(null, snapshot);
+                return null;
+              }
+            },
+            TransactionOptions.readOnlyOptionsBuilder().build());
+
+    runTransaction.get(10, TimeUnit.SECONDS);
+    assertEquals("bar", ref.get().get("foo"));
+  }
+
+  @Test
+  public void readOnlyTransaction_faliureWhenAttemptingWrite()
+      throws InterruptedException, TimeoutException {
+
+    final DocumentReference documentReference = randomColl.document("tx/ro/writeShouldFail");
+    final ApiFuture<Void> runTransaction =
+        firestore.runTransaction(
+            new Function<Void>() {
+              @Override
+              public Void updateCallback(Transaction transaction) {
+                transaction.set(documentReference, SINGLE_FIELD_MAP);
+                return null;
+              }
+            },
+            TransactionOptions.readOnlyOptionsBuilder().build());
+
+    try {
+      runTransaction.get(10, TimeUnit.SECONDS);
+    } catch (ExecutionException e) {
+      final Throwable cause = e.getCause();
+      assertThat(cause).isInstanceOf(FirestoreException.class);
+      final Throwable rootCause = ExceptionUtils.getRootCause(cause);
+      assertThat(rootCause).isInstanceOf(StatusRuntimeException.class);
+      final StatusRuntimeException invalidArgument = (StatusRuntimeException) rootCause;
+      final Status status = invalidArgument.getStatus();
+      assertThat(status.getCode()).isEqualTo(Code.INVALID_ARGUMENT);
+      assertThat(status.getDescription()).contains("read-only");
+    }
+  }
+
+  @Test
+  public void readOnlyTransaction_failureWhenAttemptReadOlderThan60Seconds()
+      throws ExecutionException, InterruptedException, TimeoutException {
+    final DocumentReference documentReference = randomColl.add(SINGLE_FIELD_MAP).get();
+
+    final TransactionOptions options =
+        TransactionOptions.readOnlyOptionsBuilder()
+            .setReadTime(com.google.protobuf.Timestamp.newBuilder().setSeconds(1).setNanos(0))
+            .build();
+
+    final ApiFuture<Void> runTransaction =
+        firestore.runTransaction(
+            new Function<Void>() {
+              @Override
+              public Void updateCallback(Transaction transaction) throws Exception {
+                transaction.get(documentReference).get(5, TimeUnit.SECONDS);
+                return null;
+              }
+            },
+            options);
+
+    try {
+      runTransaction.get(10, TimeUnit.SECONDS);
+    } catch (ExecutionException e) {
+      final Throwable rootCause = ExceptionUtils.getRootCause(e);
+      assertThat(rootCause).isInstanceOf(StatusRuntimeException.class);
+      final StatusRuntimeException invalidArgument = (StatusRuntimeException) rootCause;
+      final Status status = invalidArgument.getStatus();
+      assertThat(status.getCode()).isEqualTo(Code.FAILED_PRECONDITION);
+      assertThat(status.getDescription()).contains("old");
```

Comment:
Did you happen to run these tests against the Emulator as well? It sometimes returns slightly different error messages.

Reply:
Yup, ran them against the emulator and they all passed.